### PR TITLE
Limit the number of queued dagruns created by the Scheduler

### PR DESCRIPTION
**Pull request by ephraimbuddy**
There's no limit to the amount of queued dagruns to create currently
and it has become a concern with issues raised against it. See #18023 and #17979

This PR adds a limit to queued dagruns a scheduler can create per dag

cc: @SamWheating 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py
Comment (SamWheating) [inline airflow/jobs/scheduler_job.py:844]:
```suggestion
            if total_queued >= max_queued_dagruns:
				self.log.debug("DAG %s is at max_queued_runs_per_dag (%d), not creating any more queued runs", dag_model.dag_id, max_queued_dagruns)
                continue
```

I'm not sure if this will get too noisy, but it might be worthwhile to have some logging here.

File: airflow/config_templates/config.yml
Comment (SamWheating) [inline airflow/config_templates/config.yml]:
```suggestion
        The maximum number of queued dagruns for a single DAG. The scheduler will not create more DAG runs
```

File: airflow/jobs/scheduler_job.py
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:844]: It will get noisy. Imagine possible dag_run to create is up to 2000, this would print ~1984 times

**Comment by ashb** [comment]
I wonder if it's worth adding another conditional index on queued here like we did in https://github.com/apache/airflow/pull/18061

(Annoyingly for Mysql we'd _like_ only a single (state,dag_id) index, but I'm not sure if we can easily/realistically do that.)

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:844]: Debug is off by default, so we could have this on only at debug level without overwhelming people in normal cases.

**Comment by ephraimbuddy** [comment]
> I wonder if it's worth adding another conditional index on queued here like we did in #18061
> 
> (Annoyingly for Mysql we'd _like_ only a single (state,dag_id) index, but I'm not sure if we can easily/realistically do that.)

I ran airflow with having two indexes on (state,dag_id) in MySQL 5.7, it didn't complain. Will still test with other versions or what do you think?

**Comment by ashb** [comment]
> I ran airflow with having two indexes on (state,dag_id) in MySQL 5.7, it didn't complain. Will still test with other versions or what do you think?

Two indicies isn't a _problem_, it's just wasteful (time to udpate both, space for storing it) is all. But it's not worth spending much time on it I don't think.

**Comment by ephraimbuddy** [comment]
> > I ran airflow with having two indexes on (state,dag_id) in MySQL 5.7, it didn't complain. Will still test with other versions or what do you think?
> 
> Two indicies isn't a _problem_, it's just wasteful (time to udpate both, space for storing it) is all. But it's not worth spending much time on it I don't think.

Ok. I will remove the index now

**Comment by ashb** [comment]
I think we kind of want it for MySQL is the issue, but from what I can see in SQLA there isn't a way to define an index for certain DB engines only.

File: airflow/models/dagrun.py
Comment (ephraimbuddy) [inline airflow/models/dagrun.py:116]: This PR now depends on #18061 because of this comment. Will merge after merging #18061

**Review by ashb — APPROVED** [review]

**Pull request merged**
